### DISCUSSION TITLE
[亮度] 在DDE允许用户调节真实的背光亮度

### DISCUSSION THREAD

**Discussion by foresee-io** [linuxdeepin/developer-center#3370] [Features Request & Ideas | 特性请求 & 头脑风暴]
2022/12/19: 增加“用命令操控真实背光亮度”栏目

## 背景

NVIDIA独显直通时，由于默认设置，背光亮度是不可调的。 **截至 2022/12/19 deepin v20.8** ，DDE采用了变通方法调背光，可能是加一层遮罩或调Gamma？
但毕竟 **调节的并非真实背光**，低“亮度”下LCD屏的观感会打折扣。

题外话：[在dock栏直接调亮度](https://github.com/linuxdeepin/developer-center/discussions/3253)的想法不错！

---

### 0. 解决N卡直通时，背光不可调问题
一般地，在 Debian 系的系统运行下列命令后，重启生效
``` bash
sudo cat << EOF >> /etc/modprobe.d/custom.conf
options nvidia NVreg_RegistryDwords=EnableBrightnessControl=1
options nvidia-current NVreg_RegistryDwords=EnableBrightnessControl=1
EOF
sudo update-initramfs -u
```

### 1. 用命令操控真实背光亮度
GNOME/KDE/XFCE/... 的亮度滑块调的是真实背光亮度，不影响LCD屏的“通透感”。
但是DDE控制中心用的是变通方法调亮度，那我们就换个办法调真实亮度。
``` bash
# i代表背光亮度（0~100）；设备路径要结合自己的机器更改
i=70 ;sudo sh -c "echo $i > /sys/devices/pci0000:00/0000:00:01.0/0000:01:00.0/backlight/nvidia_0/brightness"
```

### 建议

在“控制中心”或dconf里增添一开关，允许用户选择 **是否** 以“兼容模式”（DDE目前的方式）调节亮度。
若 **不** 以“兼容模式”调亮度，则可直接控制ACPI Backlight，以期LCD屏在低亮度下获得更好观感。

_Keyword: brightness, backlight, ACPI, 亮度, 背光_

**Comment by foresee-io** [comment]
还是先问问“有没有”。目前deepin有没有上文提到的开关 or 配置文件呢

**Reply by zccrs** [reply]
这样的配置据我所知还没有。不过dde是直接直接调设备的背光的，笔记本的屏幕基本上都支持。不支持的情况会回退到调整gamma，我忘记这个判断条件是怎么写的了，明天扒代码看一下。

**Comment by foresee-io** [comment]
22/8/11
## 判断方法
``` bash
# 方案仅针对本案例、本电脑，未必通用
# _path变量需根据实际设备调整
_path="/sys/devices/pci0000:00/0000:00:01.0/0000:01:00.0/backlight/nvidia_0/brightness"
# 显示 100 ，说明DDE没有调真实背光
cat "$_path"
# 验证真实背光可调，前提是配好nvidia模块参数
sudo sh -c "echo 100 >$_path;echo 50 >$_path;sleep 5s;echo 100 >$_path"
```

## 硬件信息
- 型号: Lenovo 81T0 (拯救者Y7000 2019 PG0)
- 图形: NVIDIA GeForce GTX 1650 (独显直通)
  - 驱动: 从官方 `non-free` 源安装 `nvidia-driver` ，版本 510.60.02
 - 屏幕: AU Optronics Corporation (DP-4)
- 启动: UEFI

**Reply by zccrs** [reply]
可以加支持，如果有代码提交，到时候麻烦你帮忙测一下

**Comment by xzl01** [comment]
这个问题我们当时考虑过，新的dtk模块就准备采用这种方式进行调整，但是在预先调研的过程中，发现这个问题不仅仅是需要显卡支持还需要内核的支持。我看到linux内核那边红帽的工程师正在解决这个问题，现有的方案调整受限于特定显卡和显示器。这个会持续跟进解决的
